[service-bus] Track2 samples and readme update (#7997)
Azure/azure-sdk-for-js: sdk/servicebus/service-bus/README.md

- Send messages to a Queue or Topic
- Receive messages from a Queue or Subscription

[Source code](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/servicebus/service-bus) | [Package (npm)](https://www.npmjs.com/package/@azure/service-bus) | [API Reference Documentation](https://azuresdkdocs.blob.core.windows.net/$web/javascript/azure-service-bus/5.0.0-preview.1/index.html) | [Product documentation](https://azure.microsoft.com/en-us/services/service-bus/) | [Samples](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/servicebus/service-bus/samples)

**NOTE**: If you are using version 1.1.x or lower, then please use the below links instead

[Source code for v1.1.5](https://github.com/Azure/azure-sdk-for-js/tree/%40azure/service-bus_1.1.5/sdk/servicebus/service-bus) |
[Package for v1.1.5 (npm)](https://www.npmjs.com/package/@azure/service-bus/v/1.1.5) |
[Samples for v1.1.5](https://github.com/Azure/azure-sdk-for-js/tree/%40azure/service-bus_1.1.5/sdk/servicebus/service-bus/samples)

## Getting Started

### Install the package

Install the Azure Service Bus client library using npm

> `npm install @azure/service-bus@dev`

Reviewer: @richardpark-msft This is still pointing to dev instead of next ...

Author: Blech, PR out to fix it: https://github.com/Azure/azure-sdk-for-js/pull/8056

Install the Azure Service Bus client library using npm

> `npm install @azure/service-bus@dev`

Reviewer 2: Nice, did not know that we have a dev tag.

But I think we have to do `2.0.0-preview.1` or `next`.

Reviewer: Harsha is right
`dev` tag would get the nightly builds that we have.

Also, I want to discuss options to ensure that a user looking for the stable build doesnt end up using the samples here and is re-directed properly. My previous idea was to add the note that you see in line 12, but now I feel we should have something more prominent. And also maybe at the start of every H3 or H2 just cause people might sometimes scroll to the header they are interested and skip the top part.

Thoughts?

Author: Good thoughts (from both of you!).

- I can definitely reference preview.1 explicitly. We're just in that annoying state that none of the bits are "published" under a nice tag yet :)
- For the "user looking for samples" we have (AFAIK) two avenues they could find these:
  - Via docs.microsoft.com : @willmtemple recently removed the "front matter" that would cause docs.microsoft.com to index these samples. So at least there, only the stable samples will show up.
  - Via github: this one is problematic. @HarshaNalluru was thinking we could just make a separate folder so we could at least host them side-by-side. @willmtemple , have we adopted any sort of convention around this yet (preview and stable hosted side-by-side?)

Reviewer 2: Ah.. Thanks for asking, this reminded me that I wanted to mention regarding samples-v1 here. 

For now, please keep the copy of samples for version-1 here in a separate folder `samples-v1` until we have the plan to manage both the versions.

Author: @HarshaNalluru  - I'm going to link to them via the git tag - is that not enough? Why do they need to be in master?

Author: After talking with @HarshaNalluru  I believe we're good with the approach we've got here.

Reviewer 2: Okay, only keeping v2-preview-version, deleting the v1-samples is fine.

Author: @ramya-rao-a  - I've now inserted a comment into:

- Each source file (.js and .ts)
- The readmes for each folder

So hopefully that's enough of a breadcrumb :)

### Authenticate the client

> Interaction with Service Bus starts with an instance of the [ServiceBusClient](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/servicebusclient) class.

Reviewer: These links will go to Track 1. And some of them might be 404 until we publish. Others will still go 404 when are in Preview 2 and change APIs

In Event Hubs, we used the azure.github.io links while we were in preview and moved to the docs.microsoft.com links only after the APIs settled down.

Author: It was also super annoying because I also had to guess I'll swap over to github.io and kill two birds with one stone. :)

Author: Swapped links over to what I think the azuresdkdocs.blob.core.windows.net proper URLs will be.

You can instantiate this class using its constructors:

- [Create using a connection string](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/servicebusclient#servicebusclient-string--servicebusclientoptions-)
  - This method takes the connection string to your Service Bus instance. You can get the connection string
    from the Azure portal.
- [Create using a TokenCredential](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/servicebusclient#servicebusclient-string--tokencredential---servicebusclientoptions-)
  - This method takes the host name of your Service Bus instance and a credentials object that you need
    to generate using the [@azure/identity](https://www.npmjs.com/package/@azure/identity)

>     library. The host name is of the format `name-of-service-bus-instance.servicebus.windows.net`.

Reviewer 2: Host name doesn't include `sb://` ?

Reviewer: I dont think so..

Author: Not in these cases. It does when you use a connection string but this is just the FQDN of the service bus instance.

### Key concepts

Once you have initialized the [ServiceBusClient](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/servicebusclient)

> class, use the below methods to create client objects for Queues, Topics and Subscriptions

Reviewer: We no longer have "client objects"...
Previously these were meant to be the QueueClient, TopicClient and SubscriptionClient

Reviewer: ```suggestion
class, create a sender or receiver based on whether you want to send or receive messages.
```

### Send messages

Once you have created an instance of a `ServiceBusClient` class, you can get a `Sender`
using the [getSender](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/servicebusclient#getsender-string-)

> function. This gives you a sender which you can use to [send](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/sender#send-servicebusmessage-)

Reviewer: Use the term `method` when the function is on a class. Use `function` when it is a standalone function

Author: ```suggestion
method. This gives you a sender which you can use to [send](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/sender#send-servicebusmessage-)
```

> messages.

Reviewer 2: Breaking in between the lines?
(Same at many other places.)

Author: These represent my struggle between wanting to follow some sort of reasonable "make this look nice if you're just viewing the markdown and not a rendering of the markdown". I can remove it but that's why I keep doing it :)

Author: Fixed - it actually made a nice sentence on it's own so it was win-win.

You can also use the [sendBatch](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/sender#sendbatch-servicebusmessagebatch-)

> method to send multiple messages using a single call.

Reviewer: We now need to create a batch, fill it in and then send it. Its no longer a single call.

Author: ```suggestion
method to efficiently send multiple messages in a single send.
```

const sender = serviceBusClient.getSender("my-queue");
await sender.send({
  body: "my-message-body"
});

const batch = await sender.createBatch();

// NOTE: tryAdd() returns false if the message could not
// be added because the batch is at capacity.
batch.tryAdd({ body: "my-message-body-1" });
batch.tryAdd({ body: "my-message-body-2" });
batch.tryAdd({ body: "my-message-body-3" });

await sender.sendBatch(batch);
```

### Receive messages

Once you have created an instance of a `ServiceBusClient` class, you can get a `Receiver`
using the [getReceiver](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/servicebusclient#getreceiver-string--string-) function.

```javascript
const receiver = serviceBusClient.getReceiver("my-queue", "peekLock");
```

You can use this receiver in one of 3 ways to receive messages:

#### Get an array of messages

Use the [receiveBatch](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/receiver#receivebatch-number--receivebatchoptions-) function which returns a promise that
resolves to an array of messages.

```javascript
const myMessages = await receiver.receiveBatch(10);
```

#### Subscribe using a message handler

> Use the [subscribe](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/receiver#subscribe-messagehandlers--subscribeoptions-)

Reviewer 2: Use the subscribe -> Use the subscribe method?

Author: Hahah, yes. :|

Author: ```suggestion
Use the [subscribe](https://docs.microsoft.com/en-us/javascript/api/%40azure/service-bus/receiver#subscribe-messagehandlers--subscribeoptions-) method
```
